### ISSUE TITLE
[BUG] coregistration issues

### ISSUE THREAD

**Issue by mattcieslak** [PennLINC/qsiprep#740]
## Description

@hamsiradhakrishnan and @cookpa found that in certain cases the b=0 to anatomical coregistration is not accurate. This issue was introduced when the best-b=0 method was added in 0.16. Coregistration was being performed with the undistorted output from TOPUP, not with the undistorted reference used by Eddy. This could result in a bad coregistration if the TOPUP reference is not closely aligned with the reference used by Eddy.

The problem is fixed in #739 and will be part of release 0.21.3

## Severity

In most cases this bug will have no effect - the first b=0 tends to be selected for TOPUP or a b=0 that is closely aligned to it. If the first b=0 is selected for TOPUP then this bug will have no effect at all. Otherwise there will be some degree of misalignment. You would see the misalignment in the final data, not in the html reports. 

The fieldmap is correctly registered and applied to the data using --field and --field_mat. 

As far as recon workflows go, this can affect any method that uses both anatomical _and_ dwi data (eg connectivity matrices).

## Summary

In light of this bug and the other bugs fixed in 0.21.3 we recommend re-running your data with 0.22 when it is released. If you have noisy data with many b=0s in it, you may be impacted by this bug.

**Comment by mattcieslak** [comment]
Will reopen this if problems are reported with v 0.21.3

**Comment by cookpa** [comment]
Related to this, I took a look at the topup selection. The qc_score and qc_rank appear to be uncorrelated. Here's the file

`/workdir/qsiprep_wf/single_subject_REDACTED_wf/dwi_preproc_ses_REDACTED_acq_34dir_wf/hmc_sdc_wf/ds_topupcsv/qsiprep/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_topupcsv.csv`

```
bids_origin_file,original_volume,fsl_spec,qc_score,qc_rank,selected_for_sdc
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,0,0 -1 0 0.030904,-0.7488178704289179,7,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,1,0 -1 0 0.030904,-0.7530345706945659,5,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,10,0 -1 0 0.030904,-0.7567398646702149,6,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,19,0 -1 0 0.030904,-0.7549840038890062,13,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,28,0 -1 0 0.030904,-0.7583652438242636,9,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,35,0 -1 0 0.030904,-0.765407356215387,8,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,36,0 -1 0 0.030904,-0.7630675256138219,4,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,45,0 -1 0 0.030904,-0.7658788440591788,2,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,54,0 -1 0 0.030904,-0.758565784230171,3,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,63,0 -1 0 0.030904,-0.7589432484036898,12,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,70,0 -1 0 0.030904,-0.7521138061581196,11,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,71,0 -1 0 0.030904,-0.7532746817709385,1,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,80,0 -1 0 0.030904,-0.7549074644782986,10,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,89,0 -1 0 0.030904,-0.761922922990584,14,False
/data/input/sub-REDACTED/ses-REDACTED/dwi/sub-REDACTED_ses-REDACTED_acq-34dir_dwi.nii.gz,98,0 -1 0 0.030904,-0.7495245533109298,0,True
/data/input/sub-REDACTED/ses-REDACTED/fmap/sub-REDACTED_ses-REDACTED_acq-synb0_dir-PA_epi.nii.gz,0,0 1 0 0.000000,2.0,0,True
```

**Comment by cookpa** [comment]
This is with 0.21.3, btw

**Comment by mattcieslak** [comment]
I wonder if the ordering is off when adding the columns to the main df from the groupby df:
```
    spec_groups = dwi_b0_df.groupby("fsl_spec")
    max_per_spec = min(max_per_spec, min(spec_groups.apply(len)))

    # Calculate the "quality" of each image:
    dwi_b0_df["qc_score"] = spec_groups["nii_3d_files"].transform(calculate_best_b0s)
    dwi_b0_df["qc_rank"] = spec_groups["qc_score"].transform(np.argsort)
```

**Comment by mattcieslak** [comment]
I have no idea what argsort is doing in that transform call. Fixed in #744
